Use Docker containers in the RTW

alistairsellar opened the pull request:
<!--
    Thank you for contributing to our project!

    Please do not delete this text completely, but read the text below and keep
    items that seem relevant. If in doubt, just keep everything and add your
    own text at the top, a reviewer will update the checklist for you.

    While the checklist is intended to be filled in by the technical and scientific
    reviewers, it is the responsibility of the author of the pull request to make
    sure all items on it are properly implemented.
-->

## Description

<!--
    Please describe your changes here, especially focusing on why this pull request makes
    ESMValTool better and what problem it solves.

    Before you start, please read our contribution guidelines: https://docs.esmvaltool.org/en/latest/community/

    Please fill in the GitHub issue that is closed by this pull request, e.g. Closes #1903
-->
Our strategy for the recipe test workflow is to use containers, in order to maximise consistency of environment. This PR will implement use of containers on Jasmin, taking `latest` container from dockerhub and running with singularity.

- Closes #3246 
- Link to documentation: TBA

* * *

## Before you get started

<!--
    Please discuss your idea with the development team before getting started,
    to avoid disappointment or unnecessary work later. The way to do this is
    to open a new issue on GitHub.
-->

- [ ] [☝ Create an issue](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#contributing-code-and-documentation) to discuss what you are going to do

## Checklist

It is the responsibility of the author to make sure the pull request is ready to review. The icons indicate whether the item will be subject to the [🛠 Technical][1] or [🧪 Scientific][2] review.

<!-- The next two lines turn the 🛠 and 🧪 below into hyperlinks -->
[1]: https://docs.esmvaltool.org/en/latest/community/review.html#technical-review
[2]: https://docs.esmvaltool.org/en/latest/community/review.html#scientific-review

- [ ] [🛠][1] This pull request has a [descriptive title](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#pull-request-title)
- [ ] [🛠][1] Code is written according to the [code quality guidelines](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#code-quality)
- [ ] [🛠][1] [Documentation](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#documentation) is available
- [ ] [🛠][1] [Tests](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#tests) run successfully
- [ ] [🛠][1] The [list of authors](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#list-of-authors) is up to date
- [ ] [🛠][1] Any changed dependencies have been [added or removed correctly](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#dependencies)
- [ ] [🛠][1] All [checks below this pull request](https://docs.esmvaltool.org/en/latest/community/code_documentation.html#pull-request-checks) were successful

### [New or updated recipe/diagnostic](https://docs.esmvaltool.org/en/latest/community/diagnostic.html)

- [ ] [🧪][2] [Recipe runs successfully](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#testing-recipes)
- [ ] [🧪][2] [Recipe is well documented](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#recipe-and-diagnostic-documentation)
- [ ] [🧪][2] [Figure(s) and data](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#diagnostic-output) look as expected from literature
- [ ] [🛠][1] [Provenance information](https://docs.esmvaltool.org/en/latest/community/diagnostic.html#recording-provenance) has been added

### [New or updated data reformatting script](https://docs.esmvaltool.org/en/latest/develop/dataset.html)

- [ ] [🛠][1] [Documentation](https://docs.esmvaltool.org/en/latest/community/dataset.html#dataset-documentation) is available
- [ ] [🛠][1] The dataset has been [added to the CMOR check recipe](https://docs.esmvaltool.org/en/latest/community/dataset.html#testing)
- [ ] [🛠][1] The dataset has been added to the shared [data pools](https://docs.esmvaltool.org/en/latest/community/dataset.html#cmorized-data) of DKRZ and Jasmin by the @ESMValGroup/OBS-maintainers team
- [ ] [🧪][2] Numbers and units of the data look [physically meaningful](https://docs.esmvaltool.org/en/latest/community/dataset.html#scientific-sanity-check)

***

To help with the number of pull requests:

-  🙏 We kindly ask you to [review](https://docs.esmvaltool.org/en/latest/community/review.html#review-of-pull-requests) two other [open pull requests](https://github.com/ESMValGroup/ESMValTool/pulls) in this repository

<!--
If you need help with any of the items on the checklists above, please do not hesitate to ask by commenting in the issue or pull request.
-->

valeriupredoi commented on esmvaltool/utils/recipe_test_workflow/recipe_test_workflow/app/build_container/rose-app.conf:
```diff
@@ -0,0 +1,10 @@
+[command]
+default=singularity build ${CONTAINER_DIR}/${CONTAINER_FILE} ${DOCKER_SOURCE}
```
just a quick heads up one would prob best check with JASMIN which `singularity` is best to be used ie the system default one is pretty old, maybe they have a dedicated setup to build singularity containers (that I am not aware of, sadly)

alistairsellar replied:
Yup, helpdesk were good with using this version of singularity. Singularity has been renamed as apptainer, but this not installed on Jasmin yet. Current installation is doing the job we need it to for now.

valeriupredoi reviewed (commented):
just a quick comment from me (since, for some reason, I am now stuck to all these RTW PRs anyway :grin: ). And another one, for good measure: would it be possible to have such dir as `bin` shipped in the package? `tmp`, `bin` - all these system-like dirs may create confusions - our package is well built as it is now, but we never know in the future

valeriupredoi commented:
@esmvalbot Please run examples/recipe_python.yml

valeriupredoi commented:
@esmvalbot Please run examples/recipe_python.yml

valeriupredoi commented:
@esmvalbot Please run examples/recipe_python.yml

valeriupredoi commented:
my apologies @alistairsellar - I was testing, and eventually brought the Bot back to life, and your PR was unlucky enough to be the first one in the list, so I hijacked it - one :beer: for you!

alistairsellar commented:
No worries @valeriupredoi , I just assumed that you were stalking my (draft) PR :o)

rbeucher commented:
Hi @alistairsellar 

It is our strategy as well. Things work well so far except for Julia. See my issue here #3410

R

valeriupredoi commented:
nobody cares about Julia, mate 🤣 That said, I'll have a look at fixing her tomorrow 👍

rbeucher commented:
Thanks 😁😉

alistairsellar commented:
Thanks for your help with this one @ehogan. It's all now working and tests passing. Successfully completes on both Met Office and Jasmin. Please can you review?

alistairsellar commented on esmvaltool/utils/recipe_test_workflow/recipe_test_workflow/app/configure/bin/configure.py:
```diff
@@ -30,8 +30,7 @@ def main():
 def get_config_values_from_task_env():
-    """Return the configuration values defined in the environment for the
-    ``configure`` task."""
+    """Get configuration values from environment for configure task."""
```
As discussed offline, I accept your offer to ignore :smiley:

alistairsellar commented on esmvaltool/utils/recipe_test_workflow/doc/source/user_guide/workflow.rst:
```diff
@@ -18,13 +18,16 @@ The |RTW| performs the following steps:
   :Executes:
-     The ``clone_latest_esmval.sh`` script from the |Rose| app
+     The ``clone_latest_esmval.sh`` script from the |Rose| appi (if cloning),
+     or ``singularity build`` (if getting container)
```
Thanks. I implemented something similar (https://github.com/ESMValGroup/ESMValTool/pull/3399/commits/61fad57f23b26fcb5b4fc26bef14d6329ed0154c), with slight tweak as there isn't a singularity build script - it's run as a command specified directly in `rose-app.conf`

alistairsellar commented on esmvaltool/utils/recipe_test_workflow/doc/source/user_guide/workflow.rst:
```diff
@@ -18,13 +18,16 @@ The |RTW| performs the following steps:
   :Description:
-     Checks out the latest versions of |ESMValTool| and |ESMValCore| from GitHub
+     Either clones the latest versions of |ESMValTool| and |ESMValCore| from GitHub, 
+     or gets the latest container from DockerHub and converts to a singularity
```
Yup. Changed in https://github.com/ESMValGroup/ESMValTool/pull/3399/commits/982ebc205d67cc6d633fc3a86e4c3567fe700486

alistairsellar commented:
> The RTW also completed successfully on JASMIN :) However, the `get_esmval` task took 28 minutes! 😮 Did I do something wrong? 🤔

Hrrrm, that's very slow. I had observed that it was often slower on LOTUS than running interactively on sci3 (the latter takes about 5 mins). I believe that there are LOTUS performance problems at present, and indeed many of the other tasks are taking longer or timing out for me today (singularity build on sci3 still takes 5 mins today). When the LOTUS problems are resolved I'll test again and open a new issue if significantly longer than 5 mins.

alistairsellar commented on esmvaltool/utils/recipe_test_workflow/recipe_test_workflow/opt/rose-suite-jasmin.conf:
```diff
@@ -1,8 +1,8 @@
-KGO_ROOT_PATH="/gws/nopw/j04/esmeval/KGO_v2.7.0/"
-MODULE_NAME="esmvaltool/2.7"
+KGO_ROOT_PATH="/gws/nopw/j04/esmeval/KGO"
+MODULE_NAME="latest"
```
Nice suggestion - I had been unsatisfied with the old name. Done in https://github.com/ESMValGroup/ESMValTool/pull/3399/commits/e8824ccf913469e3f601c55eddefc5b2d882a5ba and https://github.com/ESMValGroup/ESMValTool/pull/3399/commits/de7397e836cf4eeeeb1b7e81c7edc688db5fed7d.

alistairsellar commented:
Thanks for the review @ehogan! I believe I've addressed all the comments. The latest code runs for me on both Jasmin and Met Office sites. Are you happy with the new changes?

ehogan commented on esmvaltool/utils/recipe_test_workflow/doc/source/user_guide/workflow.rst:
```diff
@@ -18,13 +18,16 @@ The |RTW| performs the following steps:
+     image, depending on ``SITE``.
   :Runs on:
-     Localhost
+     Localhost (if cloning), or compute node (if getting container).
```
This reverted a little too much! So, as just discussed offline, I reintroduced some text in [e47a70d](https://github.com/ESMValGroup/ESMValTool/pull/3399/commits/e47a70d947bfcbe21d85c7c1cc527eae6866eed1) :)

ehogan reviewed (approved):
Nice work @alistairsellar! 🥳 As just discussed, I made a few minor changes in [e47a70d](https://github.com/ESMValGroup/ESMValTool/pull/3399/commits/e47a70d947bfcbe21d85c7c1cc527eae6866eed1).

alistairsellar commented:
Thanks @ehogan! Very good final tweaks. Many thanks for all your help with this one!
